agent/engine/task_manager_test.go

```diff
@@ -1622,6 +1623,255 @@ func TestApplyResourceStateFailures(t *testing.T) {
+		Err                error
+		ChangedKnownStatus taskresource.ResourceStatus
+		TaskDesiredStatus  apitask.TaskStatus
+	}{
+		{
+			Name:               "steady state transition already done",
+			KnownStatus:        taskresource.ResourceStatus(volume.VolumeCreated),
+			DesiredKnownStatus: taskresource.ResourceStatus(volume.VolumeCreated),
+			Err:                nil,
+			ChangedKnownStatus: taskresource.ResourceStatus(volume.VolumeCreated),
+			TaskDesiredStatus:  apitask.TaskRunning,
+		},
+		{
+			Name:               "transition state less than known status",
+			DesiredKnownStatus: taskresource.ResourceStatus(volume.VolumeStatusNone),
+			Err:                nil,
+			KnownStatus:        taskresource.ResourceStatus(volume.VolumeCreated),
+			ChangedKnownStatus: taskresource.ResourceStatus(volume.VolumeCreated),
+			TaskDesiredStatus:  apitask.TaskRunning,
+		},
+	}
+	volumeName := "vol"
+	for _, tc := range testCases {
+		t.Run(tc.Name, func(t *testing.T) {
+			res := &volume.VolumeResource{Name: volumeName}
+			res.SetKnownStatus(tc.KnownStatus)
+			mtask := managedTask{
+				Task: &apitask.Task{
+					Arn:                 "task1",
+					ResourcesMapUnsafe:  make(map[string][]taskresource.TaskResource),
+					DesiredStatusUnsafe: apitask.TaskRunning,
+				},
+			}
+			mtask.AddResource(volumeName, res)
+			mtask.handleResourceStateChange(resourceStateChange{
+				res, tc.DesiredKnownStatus, tc.Err,
+			})
+			assert.Equal(t, tc.ChangedKnownStatus, res.GetKnownStatus())
+			assert.Equal(t, tc.TaskDesiredStatus, mtask.GetDesiredStatus())
+		})
+	}
+}
+
+func TestVolumeResourceNextState(t *testing.T) {
+	testCases := []struct {
+		Name             string
+		ResKnownStatus   taskresource.ResourceStatus
+		ResDesiredStatus taskresource.ResourceStatus
+		NextState        taskresource.ResourceStatus
+		ActionRequired   bool
+	}{
+		{
+			Name:             "next state happy path",
+			ResKnownStatus:   taskresource.ResourceStatus(volume.VolumeStatusNone),
+			ResDesiredStatus: taskresource.ResourceStatus(volume.VolumeCreated),
+			NextState:        taskresource.ResourceStatus(volume.VolumeCreated),
+			ActionRequired:   true,
+		},
+		{
+			Name:             "desired terminal",
+			ResKnownStatus:   taskresource.ResourceStatus(volume.VolumeStatusNone),
+			ResDesiredStatus: taskresource.ResourceStatus(volume.VolumeRemoved),
+			NextState:        taskresource.ResourceStatus(volume.VolumeRemoved),
+			ActionRequired:   false,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.Name, func(t *testing.T) {
+			res := volume.VolumeResource{}
+			res.SetKnownStatus(tc.ResKnownStatus)
+			res.SetDesiredStatus(tc.ResDesiredStatus)
+			mtask := managedTask{
+				Task: &apitask.Task{},
+			}
+			transition := mtask.resourceNextState(&res)
+			assert.Equal(t, tc.NextState, transition.nextState)
+			assert.Equal(t, tc.ActionRequired, transition.actionRequired)
+		})
+	}
+}
+
+func TestStartVolumeResourceTransitionsHappyPath(t *testing.T) {
```

Comment:
Should we be testing non-empty/non-happy path transitions? For example, if there are timeouts waiting for transition, or Docker has an issue handling the volume?